Changes:

Fennec-meow opened the pull request:
1. The project implements page-by—page data loading for the photo feed: a. Added a structure for decoding the Unsplash JSON response - "PhotoResult" (as well as an auxiliary structure "UrlsResult" for retrieving fields from "urls"); b. Created the "Photo" structure, which will be used in the UI part of the application; c. Upon receiving a response from the network, a "Photo" object is generated for each "PhotoResult"; d. To get the date "createdAt: Date from PhotoResult\.createdAt", "ISO8601DateFormatter" was used; e. A new class "ImagesListService" was created;
f. Added the "photos: [Photo]" property to store the sequence of all photos uploaded from the network.; g. The notification "static let didChangeNotification" has been added to the "ImagesListService". = Notification.Name (rawValue: "ImagesListServiceDidChange")"; h. The "fetchPhotosNextPage()" function has been declared in the "ImagesListService" class, which loads the next page; i. The "updateTableViewAnimated()" function has been written, which will be called upon notification "ImagesListService.didChangeNotification " and will animatedly update the state of the table.
2. Implemented the functionality of likes
3. Added the option to log out of your account
4. Added error logging to the methods of the ImagesListService service

Fennec-meow committed:
Changes:
1. The project implements page-by—page data loading for the photo feed:
a. Added a structure for decoding the Unsplash JSON response - "PhotoResult" (as well as an auxiliary structure "UrlsResult" for retrieving fields from "urls");
b. Created the "Photo" structure, which will be used in the UI part of the application;
c. Upon receiving a response from the network, a "Photo" object is generated for each "PhotoResult";
d. To get the date "createdAt: Date from PhotoResult\.createdAt", "ISO8601DateFormatter" was used;
e. A new class "ImagesListService" was created;
f. Added the "photos: [Photo]" property to store the sequence of all photos uploaded from the network.;
g. The notification "static let didChangeNotification" has been added to the "ImagesListService". = Notification.Name (rawValue: "ImagesListServiceDidChange")";
h. The "fetchPhotosNextPage()" function has been declared in the "ImagesListService" class, which loads the next page;
i. The "updateTableViewAnimated()" function has been written, which will be called upon notification "ImagesListService.didChangeNotification " and will animatedly update the state of the table.
2. Implemented the functionality of likes
3. Added the option to log out of your account
4. Added error logging to the methods of the ImagesListService service

Alex-rev-dev commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,215 @@
+    // MARK: Singletone
+    
+    static let shared = ImagesListService()
+    //    private init() {}
```
Такие артефакты можно удалять, так код будет чище :)

Alex-rev-dev commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,215 @@
+            self.task = nil
+            
+            guard case .success(let photosResult) = result else {
+                // Ничего не делайте, если пользователь прокрутит страницу вниз, будет выполнен ещё один запрос
```
Лишние комментарии можно удалить

Alex-rev-dev commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -34,79 +35,167 @@ final class ImagesListViewController: UIViewController {
-                let indexPath = sender as? IndexPath
-            else {
+                let indexPath = sender as? IndexPath else {
                 assertionFailure("Invalid segue destination")
```
Лучше избегать использование assertionFailure. Вот [тут](https://stevenpcurtis.medium.com/precondition-assert-fatal-error-or-guard-in-your-swift-code-5f9297658be0) можно почитать подробнее

Alex-rev-dev commented on ImageFeed/Profile/Presentation/ProfileViewController.swift:
```diff
@@ -8,30 +8,22 @@
+    private let profileService = ProfileService.shared
+    private let splashViewController = SplashViewController()
     private let swiftKeychainWrapper = SwiftKeychainWrapper()
+    private var username: String? = String()
```
Лучше использовать nil, если это Optional, или "" если нужно пустое значение

Fennec-meow committed:
Changes:
1. Brought the date in the cell to look like in the layouts - let's change the locale to Russian;
2. Added the wrapper "DispatchQueue.main.async" so that the "photos" array is updated from the main stream;
3. The "prepare(for segue:, sender:)" method has also been changed in the "ImagesListViewController".

Alex-rev-dev commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -45,7 +46,9 @@ final class ImagesListViewController: UIViewController {
                 let viewController = segue.destination as? SingleImageViewController,
                 let indexPath = sender as? IndexPath else {
-                assertionFailure("Invalid segue destination")
+                /// Так лучше?)
```
да, супер :)

Alex-rev-dev commented on ImageFeed/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,215 @@
+                    isLiked: photoResult.likedByUser
+                )}
+            
+            self.photos += photos
```
вот тут тоже стоит явно указать, чтобы добавление фотографии в массив происходило из главного потока

Fennec-meow committed:
Changes:
1. In the "fetchPhotosNextPage()" function, I explicitly indicated that adding an item to "photos" takes place in the main thread.
